### PR TITLE
Skip rule node which contains only inherited properties for rule cache

### PR DESCRIPTION
**Pull request by upsuper**
This is one possible fix for [bug 1427681](https://bugzilla.mozilla.org/show_bug.cgi?id=1427681) which tries to skip some rule nodes when using rule cache.

Try push for correctness: https://treeherder.mozilla.org/#/jobs?repo=try&revision=74e3941e2cfc5fba4bce839f2518af8a5a8b7411

It doesn't really show much memory saving on awsy. It only shows several KB save on fresh start memory. But since conceptually it's simple, I guess it's worth taking.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19696)

### CODE REVIEW

**Comment by upsuper** [comment]
Try pushes for performance comparison:
* base: https://treeherder.mozilla.org/#/jobs?repo=try&revision=d6d0f546f678369c82c763e16490522d08da832c
* stylo-chrome: https://treeherder.mozilla.org/#/jobs?repo=try&revision=0b5bc60bad52e1d0918895bb1fa43506362f426b
* stylo-chrome with this patch: https://treeherder.mozilla.org/#/jobs?repo=try&revision=459d9763a2458605871f3c66108bff23d6346544

File: components/style/rule_cache.rs [hunk 1]
```diff
@@ -81,12 +82,43 @@ impl RuleCache {
...
+            match *node.style_source() {
+                StyleSource::Declarations(ref decls) => {
+                    let cascade_level = node.cascade_level();
+                    let decls = decls.read_with(cascade_level.guard(guards));
+                    if decls.contains_any_reset() {
+                        break;
+                    }
+                }
+                StyleSource::None => {}
+                StyleSource::Style(_) => { break; }
```
Comment (emilio) [inline components/style/rule_cache.rs]: nit: no need for braces, just `StyleSource::Style(_) => break,`.

File: components/style/rule_cache.rs [hunk 2]
```diff
@@ -102,7 +134,8 @@ impl RuleCache {
             return None;
         }
 
-        let rules = builder_with_early_props.rules.as_ref()?;
+        let rules = builder_with_early_props.rules.as_ref();
+        let rules = RuleCache::get_rule_node_for_cache(guards, rules)?;
```
Comment (emilio) [inline components/style/rule_cache.rs]: nit: You can use `Self` if you want.

File: components/style/rule_cache.rs [hunk 3]
```diff
@@ -81,12 +82,43 @@ impl RuleCache {
...
     }
 
+    /// Walk the rule tree and return a rule node for using as the key
+    /// for rule cache.
+    ///
+    /// It currently skips a rule node when it is neither from a style
+    /// rule, nor containing any declaration of reset property. We don't
+    /// skip style rule so that we don't need to walk a long way in the
+    /// worst case. Skipping declarations rule nodes should be enough
+    /// to address common cases that rule cache fails to do reasonable
```
Comment (emilio) [inline components/style/rule_cache.rs]: You mean `s/rule cache/style sharing cache`? The rule cache should hopefully get a decent sharing now :)

**Review by emilio — APPROVED** [review]
r=me if Talos / try is not mad at this.

File: components/style/rule_cache.rs (hunk 3 above)
Comment (upsuper) [reply components/style/rule_cache.rs]: I mean, the common case that rule cache failed to do without this...

Comment (emilio) [reply components/style/rule_cache.rs]: ok, fair, maybe `would fail [...] instead` instead?

**Pull request merged**